platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,231 @@
+     */
+    int IMPLICIT_REFERENCE_COUNT = 0;
+
+    /**
+     * The reference count of an object with an explicit reference.
+     */
+    int DESTROYED_REFERENCE_COUNT = -1;
+
+    /**
+     * Acquire a reservation on this object. Increments the reference count by 1.
+     *
+     * @throws com.swirlds.common.exceptions.ReferenceCountException
+     * 		if this object has been fully released and destroyed
+     */
+    void reserve();
+
+    /**
+     * Attempts to acquire a reservation on this object. If the object is destroyed, the reservation attempt will fail.
+     *
+     * @return true if a reservation was acquired.
+     */
+    boolean tryReserve();
+
+    /**
+     * <p>
+     * Release a reservation on an object. Decrements the reference count by 1. If this method releases
+     * the last reservation, then this object should be destroyed.
+     * </p>
+     *
+     * <p>
+     * Should be called exactly once for each time {@link #reserve()} is called. The exception to this rule is
+     * if this object only has an implicit reference (i.e. {@link #getReservationCount()} returns
+     * {@link #IMPLICIT_REFERENCE_COUNT}). An object has an implicit reference immediately after it is constructed
+     * but before {@link #reserve()} has been called the first time. If called with an implicit reference, this
+     * object will be destroyed.
+     * </p>
+     *
+     * @return true if this call to release() caused the object to become destroyed
+     * @throws com.swirlds.common.exceptions.ReferenceCountException
+     * 		this object has already been fully released and destroyed
+     */
+    @Override
+    boolean release();
+
+    /**
+     * Get the total number of times {@link #reserve()} has been called minus the number of times {@link #release()}
+     * has been called. Will return {@link #IMPLICIT_REFERENCE_COUNT} if {@link #reserve()} has never been called,
+     * or {@link #DESTROYED_REFERENCE_COUNT} if this object has been fully released.
+     *
+     * @return
+     */
+    int getReservationCount();
+}
+
+```
+
+We're going to rely on this mechanism to manage the number of states that are kept in memory. 
+
+**Important**: It's up to the client code to make sure that the state is released once it's no longer needed.
+
+Once the reservation count reaches 0, the state will be eligible for garbage collection.
+
+### Java classes
+
+- `com.swirlds.state.State` has the following lifecycle-related methods:
+  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
+  - `void computeHash()`: computes the hash of the immutable state.
+  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
+  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.
+
+These methods correspond to the state lifecycle as follows:
+
+[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)
```

Comment:
This diagram and the classes now seem so to be a bit contradictory. Is describes a `State` class that has methods like `copy()` and `computeHash()`, yet these operations are described/depicted as part of the lifecycle of the state. Then there is the `StateLifecycleManager` that has `copy()`. In all cases, `copy()` can only be called on the mutable state. So we need to be very clear about which operations are part of the lifecycle and put those in the lifecycle manager and not duplicate the copy method.

Reply:
I think that the hash method can only start part of the "lifecycle" if it is done automatically on the immutable state created on copy. However I don't think this will work because hashing is expensive and, at least on the consensus node, should not be performed on the same thread as the copy itself.